General API for compression codecs

idelpivnitskiy opened the pull request:
Motivation:

Each compression codec is presented as its own class. It's difficult to switch one compression codec to another one without significant code modifications.

Modifications:

* `CompressionFormat` - enum with all available compression formats;
* `CompressionDecoder` - common interface for all compression decoders;
* `CompressionEncoder` - common interface for all compression encoders;
* `ClosableCompressionDecoder` - common interface for compression decoders which could reach the end of the compression stream.
* `ClosableCompressionEncoder` - common interface for compression encoders which could be closed by request.
* New abstract classes which implement basic functionality of described interfaces.
* Extend all compression encoders and decoders to use new abstract classes.

Result:

General API for all compression codecs.

idelpivnitskiy commented:
@trustin please, review it here. PR is reopened against 4.1

normanmaurer commented:
@trustin PTAL again

trustin commented on codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecFactory.java:
```diff
@@ -0,0 +1,92 @@
+ * Creates a new {@link CompressionEncoder} or a new {@link CompressionDecoder}
+ * with specified {@link CompressionFormat}.
+ */
+public final class CompressionCodecFactory {
```
How about `CompressionCodecs`? i.e. `CompressionCodecs.newEncoder(...)`

trustin commented on codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecFactory.java:
```diff
@@ -0,0 +1,92 @@
+                return ZlibCodecFactory.newZlibEncoder(ZlibWrapper.NONE);
+            case ZLIB_OR_NONE:
+                throw new UnsupportedOperationException("In case of " + ZlibWrapper.ZLIB_OR_NONE
+                        + " an encoder can not be auto-detected");
```
Let's use `IllegalArgumentException` and use the following message:

    "format: " + format + " (" + ZlibWrapper.ZLIB_OR_NONE + " is not allowed for an encoder + ')'

trustin commented on codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecFactory.java:
```diff
@@ -0,0 +1,92 @@
+            case LZMA:
+                return new LzmaFrameEncoder();
+            default:
+                throw new IllegalArgumentException("Format " + format + " is not supported");
```
`"format: " + format + " (not supported)"`

trustin commented on codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecFactory.java:
```diff
@@ -0,0 +1,92 @@
+            case FASTLZ:
+                return new FastLzFrameDecoder();
+            default:
+                throw new IllegalArgumentException("Format " + format + " is not supported");
```
Ditto

trustin commented on codec/src/main/java/io/netty/handler/codec/compression/CompressionDecoder.java:
```diff
@@ -0,0 +1,45 @@
+
+    protected CompressionDecoder(CompressionFormat format) {
+        ObjectUtil.checkNotNull(format, "format");
+        this.format = format;
```
You can merge these two lines into one:

    this.format = checkNotNull(format, "format");

trustin commented on codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java:
```diff
@@ -0,0 +1,182 @@
+    /**
+     * Default delay time for close. See method {@link #closeDelay(int)}.
+     */
+    public static final int DEFAULT_CLOSE_DELAY = 10;
```
Let's make this private just in case we use an interesting strategy that does not require this property at all.

trustin commented on codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java:
```diff
@@ -0,0 +1,182 @@
+     *
+     * The default value is {@link #DEFAULT_CLOSE_DELAY}.
+     *
+     * @param delay time in seconds.
```
- Could we make this millis?
- Could we rename this property to something better? How about `lastWriteTimeoutMillis`?

trustin commented on codec/src/main/java/io/netty/handler/codec/compression/CompressionFormat.java:
```diff
@@ -0,0 +1,32 @@
+    LZF,
+    LZ4,
+    FASTLZ,
+    LZMA
```
Please add Javadoc for each enum values with a href links to its web site (or wikipedia page).

trustin commented on codec/src/main/java/io/netty/handler/codec/compression/FastLzFrameEncoder.java:
```diff
@@ -183,4 +185,29 @@ protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) throws
+
+    @Override
+    protected ChannelFuture finishEncode(ChannelHandlerContext ctx, ChannelPromise promise) {
+        return promise.setFailure(new UnsupportedOperationException());
```
Instead of rejecting the request, we could simply close the channel immediately without scheduling any write timeout, etc?

idelpivnitskiy replied:
@trustin do you mean just to call `ctx.close(promise);`?

idelpivnitskiy replied:
what I should return in `boolean isClosed()` method in this case?

trustin commented on codec/src/main/java/io/netty/handler/codec/compression/LzfEncoder.java:
```diff
@@ -137,4 +138,29 @@ protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) throws
+
+    @Override
+    protected ChannelFuture finishEncode(ChannelHandlerContext ctx, ChannelPromise promise) {
+        return promise.setFailure(new UnsupportedOperationException());
```
Instead of rejecting the close request, we could simply close the channel immediately without scheduling any write timeout, etc?

trustin commented on codec/src/main/java/io/netty/handler/codec/compression/LzmaFrameEncoder.java:
```diff
@@ -212,4 +214,29 @@ private static int maxOutputBufferLength(int inputLength) {
+
+    @Override
+    protected ChannelFuture finishEncode(ChannelHandlerContext ctx, ChannelPromise promise) {
+        return promise.setFailure(new UnsupportedOperationException());
```
Instead of rejecting the close request, we could simply close the channel immediately without scheduling any write timeout, etc?

trustin commented on codec/src/main/java/io/netty/handler/codec/compression/SnappyFrameEncoder.java:
```diff
@@ -120,4 +125,29 @@ private static void writeChunkLength(ByteBuf out, int chunkLength) {
+
+    @Override
+    protected ChannelFuture finishEncode(ChannelHandlerContext ctx, ChannelPromise promise) {
+        return promise.setFailure(new UnsupportedOperationException());
```
Instead of rejecting the close request, we could simply close the channel immediately without scheduling any write timeout, etc?

trustin commented on codec/src/main/java/io/netty/handler/codec/compression/package-info.java:
```diff
@@ -15,9 +15,16 @@
+ *      <a href="https://github.com/ning/compress/wiki">LZF</a><br>
+ *      <a href="http://en.wikipedia.org/wiki/LZ4_%28compression_algorithm%29">LZ4</a><br>
+ *      <a href="http://fastlz.org/">FastLZ</a><br>
+ *      <a href="http://en.wikipedia.org/wiki/Lempel%E2%80%93Ziv%E2%80%93Markov_chain_algorithm">LZMA</a><br>
```
Could we use `ul/li` tags?

trustin commented on codec/src/main/java/io/netty/handler/codec/compression/ZlibWrapper.java:
```diff
@@ -23,18 +23,43 @@
+    /**
+     * Returns a {@link CompressionFormat} of current {@link ZlibWrapper}
+     */
+    public abstract CompressionFormat asCompressionFormat();
```
You can just add a final field to this enum and pass it to an enum constructor. No need to declare this abstract.

trustin commented:
Design comment: I see two types of encoders here:

- the one that requires finishEncode() and thus close() has to be intercepted to write the last data and its write timeout has to be scheduled.
- the one that dos not require finishEncode() and thus close() can simply close the channel immediately.

Perhaps we could split `CompressionEncoder` into three? e.g.

- CompressionEncoder (no finishEncode())
- SomeCompressionEncoder extends CompressionEncoder (overrides close() and adds finishEncode())

I'm not sure what would be a good name for SomeCompressionEncoder though. Ideas?

trustin commented:
@idelpivnitskiy Thanks a lot for your patience! Let me look forward to your update!

trustin commented:
@idelpivnitskiy Sorry I'm late. It seems like you had a few questions about my comment. What I'm suggesting is:

```java
public interface CompressionEncoder extends ChannelOutboundHandler {
    CompressionFormat format();
}

public interface ClosableCompressionEncoder extends CompressionEncoder {
    boolean isClosed();
    ChannelFuture close();
    ChannelFuture close(ChannelPromise promise);
}

public abstract class AbstractCompressionEncoder extends MessageToByteEncoder<ByteBuf> implements CompressionEncoder {
    protected SimpleCompressionEncoder(CompressionFormat format, boolean preferDirect) { ... }
    public final CompressionFormat format() { return format }
}

public class FastLzFrameEncoder extends AbstractCompressionEncoder { .. }

public abstract class AbstractClosableCompressionEncoder extends AbstractCompressionEncoder {
    ...
    public boolean isClosed() { ... }
    public ChannelFuture close(ChannelPromise promise) { ... }
    protected abstract void finishEncode(...);
}

public class Lz4FrameEncoder extends AbstractClosableCompressionEncoder { ... }
```

trustin commented:
Another feature we might need to think about is to terminate the compression stream without closing a connection. For example, a user might want to add a compression encoder to a pipeline and then remove it from the pipeline later dynamically. To achieve this, we will have to implement `handlerAdded` and `handlerRemoved`.

An advantage of this feature is that we do not need to provide `close()` or `isClosed()` method at all, because a user can simply remove the handler from the pipeline. i.e. no need for `ClosableCompressionEncoder`

trustin commented:
@idelpivnitskiy Gentle ping :-)

idelpivnitskiy commented:
@trustin thanks for the reminder. I'm a little bit busy, will look at during the next 2 weeks

normanmaurer commented:
@idelpivnitskiy any updates ?

idelpivnitskiy commented:
@normanmaurer yes, will be done this weekend

normanmaurer commented:
@idelpivnitskiy ok cool as I aim for the final release next week :)

idelpivnitskiy commented on codec/src/main/java/io/netty/handler/codec/compression/AbstractClosableCompressionEncoder.java:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    @Override
+    public final void close(final ChannelHandlerContext ctx, final ChannelPromise promise) throws Exception {
```
Is it ok that all of these three methods are final?

trustin replied:
Yep.

idelpivnitskiy commented on codec/src/main/java/io/netty/handler/codec/compression/Bzip2Encoder.java:
```diff
@@ -165,66 +155,13 @@ private void closeBlock(ByteBuf out) {
-    }
-
-    private ChannelFuture finishEncode(final ChannelHandlerContext ctx, ChannelPromise promise) {
+    protected ChannelFuture finishEncode(final ChannelHandlerContext ctx, ChannelPromise promise) {
```
Should it be `final`?

trustin replied:
Yes, please. When in doubt, add `final` and remove it later when requested with a good reason. :-)

idelpivnitskiy commented on codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecs.java:
```diff
@@ -0,0 +1,98 @@
+     *
+     * @throws IllegalArgumentException if specified format is not supported.
+     */
+    public static CompressionEncoder newEncoder(CompressionFormat format) {
```
Is it ok that factory returns high level interface? User has to cast it to `ClosableCompressionEncoder` if he wants to use additional methods.

trustin replied:
We could use generics here:

```java
public static <T extends CompressionEncoder> T newEncoder(...) {
    ...
    return (T) ...
}
```

idelpivnitskiy commented on codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecs.java:
```diff
@@ -0,0 +1,98 @@
+            case LZMA:
+                return new LzmaFrameEncoder();
+            default:
+                throw new IllegalArgumentException("format: " + format + " (not supported)");
```
Should be here something like `UnsupportedCompressionFormatException`?

trustin replied:
Yes, please.

idelpivnitskiy commented on codec/src/main/java/io/netty/handler/codec/compression/AbstractClosableCompressionEncoder.java:
```diff
@@ -0,0 +1,170 @@
+     */
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        this.ctx = null;
```
Am I right that we also have to discard link to current context if handler removes from pipeline?

idelpivnitskiy replied:
Should we call `finishEncode(...)` here?

netkins commented on codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecs.java:
```diff
@@ -0,0 +1,98 @@
+            case ZLIB:
+                return ZlibCodecFactory.newZlibEncoder(ZlibWrapper.ZLIB);
+            case GZIP:
+                return ZlibCodecFactory.newZlibEncoder(ZlibWrapper.GZIP);
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "format: " 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1192)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "format: " 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1192)

netkins commented on codec/src/main/java/io/netty/handler/codec/compression/AbstractCompressionEncoder.java:
```diff
@@ -0,0 +1,64 @@
+package io.netty.handler.codec.compression;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelFuture;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ChannelFuture'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ChannelFuture'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on codec/src/main/java/io/netty/handler/codec/compression/AbstractCompressionEncoder.java:
```diff
@@ -0,0 +1,64 @@
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ChannelHandlerContext'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ChannelHandlerContext'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on codec/src/main/java/io/netty/handler/codec/compression/AbstractCompressionEncoder.java:
```diff
@@ -0,0 +1,64 @@
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ChannelFutureListener'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ChannelFutureListener'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on codec/src/main/java/io/netty/handler/codec/compression/AbstractCompressionEncoder.java:
```diff
@@ -0,0 +1,64 @@
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.ChannelPromise;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ChannelPromise'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ChannelPromise'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on codec/src/main/java/io/netty/handler/codec/compression/AbstractCompressionEncoder.java:
```diff
@@ -0,0 +1,64 @@
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ChannelPipeline'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ChannelPipeline'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on codec/src/main/java/io/netty/handler/codec/compression/AbstractCompressionEncoder.java:
```diff
@@ -0,0 +1,64 @@
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.ChannelPromiseNotifier;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ChannelPromiseNotifier'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.channel.ChannelPromiseNotifier'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on codec/src/main/java/io/netty/handler/codec/compression/AbstractCompressionEncoder.java:
```diff
@@ -0,0 +1,64 @@
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.ChannelPromiseNotifier;
+import io.netty.handler.codec.MessageToByteEncoder;
+import io.netty.util.concurrent.EventExecutor;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.util.concurrent.EventExecutor'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.util.concurrent.EventExecutor'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on codec/src/main/java/io/netty/handler/codec/compression/AbstractCompressionEncoder.java:
```diff
@@ -0,0 +1,64 @@
+import io.netty.handler.codec.MessageToByteEncoder;
+import io.netty.util.concurrent.EventExecutor;
+import io.netty.util.internal.ObjectUtil;
+import io.netty.util.internal.OneTimeTask;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.util.internal.OneTimeTask'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.util.internal.OneTimeTask'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on codec/src/main/java/io/netty/handler/codec/compression/AbstractCompressionEncoder.java:
```diff
@@ -0,0 +1,64 @@
+import io.netty.util.internal.ObjectUtil;
+import io.netty.util.internal.OneTimeTask;
+
+import java.util.concurrent.TimeUnit;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'java.util.concurrent.TimeUnit'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'java.util.concurrent.TimeUnit'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecs.java:
```diff
@@ -0,0 +1,98 @@
+            case DEFLATE:
+                return ZlibCodecFactory.newZlibEncoder(ZlibWrapper.NONE);
+            case ZLIB_OR_NONE:
+                throw new IllegalArgumentException("format: " + format
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "format: " 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1192)

idelpivnitskiy commented on codec/src/main/java/io/netty/handler/codec/compression/AbstractClosableCompressionDecoder.java:
```diff
@@ -0,0 +1,45 @@
+    protected void handlerRemoved0(ChannelHandlerContext ctx) throws Exception {
+        if (!isClosed()) {
+            logger.warn("ClosableCompressionDecoder is removed from a pipeline, "
+                    + "but the end of the compressed stream has not been reached.");
```
Should here be an exception instead of log warning?

trustin replied:
I think it's not really a good idea to raise an exception in `handlerRemoved()`; it's like failing to close a file. Instead, we could make the `ClosableCompressionDecoder` close the stream automatically when it is removed from the pipeline.

trustin replied:
It seems like you already did what I suggested in `AbstractClosableCompressionEncodier.handlerRemoved()`. Perhaps you could do the same here?

idelpivnitskiy replied:
> I think it's not really a good idea to raise an exception in handlerRemoved(); it's like failing to close a file.

+1

> Instead, we could make the ClosableCompressionDecoder close the stream automatically when it is removed from the pipeline.

We can not close it, because decoder's state/behavior is determined by inbound compressed stream. And here we have 2 different behaviors:

1.`ClosableCompressionDecoder`s (Bzip2Decoder, Lz4FrameDecoder, JdkZlibDecoder, JZlibDecoder). All of them have similar structure of compressed stream, something like:

| stream-header | block 1 | ... | block n | stream-footer (checksum) |
|----|----|----|----|----|

So, if the user will remove handler during/after decompression of block m (1 <= m < n), and add it again after a few minutes, the stream will be corrupted. Even if the user will add it after a few completed blocks (not during the block) the final checksum in footer will rise an exception.
So, I think that user should not remove handled of `ClosableCompressionDecoder` while it's not already closed. And we may log about it or prevent.

Also notice, that all `ClosableCompressionDecoder`s are not reusable. When it's closed, all new bytes in the stream will be skipped. We may consider a new `reset` method for manual resetting of decoder (and we will use it in `handlerAdded` or doc that user have to call it manually before adding to a pipeline) and constructor parameter (or setter) to allow codec to call `reset` automatically at the end of current compressed stream.

2.`CompressionDecoder`s (LzfDecoder, FastLzFrameDecoder, SnappyFrameDecoder) are all resettable by design and decompresses each new chunk of compressed data again and again. User can easily remove it from pipeline, but decoder may be in illegal state (for example, it's already decoded header but still wait a payload). When the user adds it again to pipeline, stream will be corrupted. In this case, a new `reset` method may be helpful too.

idelpivnitskiy commented:
> Another feature we might need to think about is to terminate the compression stream without closing a connection. For example, a user might want to add a compression encoder to a pipeline and then remove it from the pipeline later dynamically. To achieve this, we will have to implement `handlerAdded` and `handlerRemoved`.

> An advantage of this feature is that we do not need to provide `close()` or `isClosed()` method at all, because a user can simply remove the handler from the pipeline. i.e. no need for `ClosableCompressionEncoder`.

@trustin I think that we shouldn't remove these methods for backward compatibility. I just added call of `finishEncode` method in `handlerRemoved` for each `ClosableCompressionEncoder`, as additional feature.
Also any encoder which does not extend `ClosableCompressionEncoder` could be easily removed and added to a pipeline without additional actions in `handlerAdded` and `handlerRemoved`.

@trustin @normanmaurer But what we will do with `CompressionDecoder`s? If such decoder is removed from pipeline and added later, there will be checksum or stream errors. User can safely remove only `ClosableCompressionDecoder`s which are already closed. I added log warning in `handlerRemoved0` about it. Also I left comments to javadoc.

trustin commented:
> @trustin I think that we shouldn't remove these methods for backward compatibility. I just added call of `finishEncode` method in `handlerRemoved` for each `ClosableCompressionEncoder`, as additional feature.
Also any encoder which does not extend `ClosableCompressionEncoder` could be easily removed and added to a pipeline without additional actions in `handlerAdded` and `handlerRemoved`.

SGTM.

> @trustin @normanmaurer But what we will do with `CompressionDecoder`s? If such decoder is removed from pipeline and added later, there will be checksum or stream errors. User can safely remove only `ClosableCompressionDecoder`s which are already closed. I added log warning in `handlerRemoved0` about it. Also I left comments to javadoc.

I'm not sure this is what you are looking for, but I left a comment above: https://github.com/netty/netty/pull/4607#discussion_r58649693

trustin commented on codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecs.java:
```diff
@@ -0,0 +1,98 @@
+     *
+     * @throws IllegalArgumentException if specified format is not supported.
+     */
+    public static CompressionDecoder newDecoder(CompressionFormat format) {
```
You could use generics here as well.

idelpivnitskiy commented on codec/src/main/java/io/netty/handler/codec/compression/UnsupportedCompressionFormatException.java:
```diff
@@ -0,0 +1,31 @@
+ *
+ * @see CompressionCodecs
+ */
+public class UnsupportedCompressionFormatException extends RuntimeException {
```
Should it be a checked exception?

idelpivnitskiy commented:
Also, I think that all `state` inner variables (like this https://github.com/netty/netty/blob/4.1/codec/src/main/java/io/netty/handler/codec/compression/LzfDecoder.java#L50) have to be `volatile`. Is't it?

netkins commented on codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecs.java:
```diff
@@ -0,0 +1,98 @@
+            case SNAPPY:
+                return (T) new SnappyFrameDecoder();
+            case BZIP2:
+                return (T) new Bzip2Decoder();
```
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Impossible cast from io.netty.handler.codec.compression.Bzip2Decoder to io.netty.handler.codec.compression.CompressionEncoder in io.netty.handler.codec.compression.CompressionCodecs.newDecoder(CompressionFormat) [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=findbugs%3ABC_IMPOSSIBLE_CAST)

netkins commented on codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecs.java:
```diff
@@ -0,0 +1,98 @@
+            case BZIP2:
+                return (T) new Bzip2Decoder();
+            case LZF:
+                return (T) new LzfDecoder();
```
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Impossible cast from io.netty.handler.codec.compression.LzfDecoder to io.netty.handler.codec.compression.CompressionEncoder in io.netty.handler.codec.compression.CompressionCodecs.newDecoder(CompressionFormat) [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=findbugs%3ABC_IMPOSSIBLE_CAST)

netkins commented on codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecs.java:
```diff
@@ -0,0 +1,98 @@
+            case ZLIB_OR_NONE:
+                return (T) ZlibCodecFactory.newZlibDecoder(ZlibWrapper.ZLIB_OR_NONE);
+            case SNAPPY:
+                return (T) new SnappyFrameDecoder();
```
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Impossible cast from io.netty.handler.codec.compression.SnappyFrameDecoder to io.netty.handler.codec.compression.CompressionEncoder in io.netty.handler.codec.compression.CompressionCodecs.newDecoder(CompressionFormat) [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=findbugs%3ABC_IMPOSSIBLE_CAST)

netkins commented on codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecs.java:
```diff
@@ -0,0 +1,98 @@
+            case LZF:
+                return (T) new LzfDecoder();
+            case LZ4:
+                return (T) new Lz4FrameDecoder();
```
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Impossible cast from io.netty.handler.codec.compression.Lz4FrameDecoder to io.netty.handler.codec.compression.CompressionEncoder in io.netty.handler.codec.compression.CompressionCodecs.newDecoder(CompressionFormat) [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=findbugs%3ABC_IMPOSSIBLE_CAST)

netkins commented on codec/src/main/java/io/netty/handler/codec/compression/CompressionCodecs.java:
```diff
@@ -0,0 +1,98 @@
+            case LZ4:
+                return (T) new Lz4FrameDecoder();
+            case FASTLZ:
+                return (T) new FastLzFrameDecoder();
```
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Impossible cast from io.netty.handler.codec.compression.FastLzFrameDecoder to io.netty.handler.codec.compression.CompressionEncoder in io.netty.handler.codec.compression.CompressionCodecs.newDecoder(CompressionFormat) [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=findbugs%3ABC_IMPOSSIBLE_CAST)

idelpivnitskiy replied:
@trustin looks like @netkins don't like it...

slandelle commented:
BTW, I guess it would be interesting to check [aircompressor](https://github.com/airlift/aircompressor) as it claims to be faster than JNI wrappers.

idelpivnitskiy commented:
@slandelle Thank you!
I will look at this

idelpivnitskiy commented:
@trustin @normanmaurer I decided to remove factory class. Looks like we don't need it, cause it created codecs just with default settings and had a problems with type casting. Let's do our compression codecs like Java Collections. We will provide interfaces and user will choose which implementations he needs.

So, we have to complete discussions just in this thread: https://github.com/netty/netty/pull/4607#discussion_r58903263

netkins commented on codec/src/main/java/io/netty/handler/codec/compression/JZlibEncoder.java:
```diff
@@ -337,30 +294,7 @@ protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) throws
-    }
-
-    private ChannelFuture finishEncode(ChannelHandlerContext ctx, ChannelPromise promise) {
+    protected final ChannelFuture finishEncode(ChannelHandlerContext ctx, ChannelPromise promise) {
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add the "@Override" annotation above this method signature [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1161)

netkins replied:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add the "@Override" annotation above this method signature [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1161)

eun-ice commented:
I have a different issue with current compression codecs: They are only usable in a pipeline. If you need to use the separately (e.g. because you just want to compress/decompress the body of HttpContents) then they are pretty hard to use. HttpContentDecompressor creates an EmbeddedChannel just to decode gzip.

This PR seems like a good starting point to make the Compressor/Decompressor usable without a ChannelHandlerContext?

I would be willing to invest time to refactor all of that, but I'ld like to agree on how the API should look like. See https://github.com/netty/netty/issues/7340

idelpivnitskiy commented:
@eun-ice thanks for your message,
Exactly this API is not related to HTTP compression/decompression and doesn't make any changes for usage of ChannelHandlerContext. It makes general interfaces to use compression codecs in a ChannelPipeline and easily swap implementation if needed, without rewriting your code.

If you design a new way to use these codecs for HTTP protocol without EmbaddedChannel, it will be great! Just make sure that existing compression codecs API will continue to work. Start with ZlibDecoder/Encoder as an example of closable compression codec and LzfDecoder/Encoder as an example of non-closable compression codec. The rest of the codecs will be easy to convert to the same API.

eun-ice commented:
@idelpivnitskiy I do want to built on top of this build request.

What I am thinking about:

Introduce an interface CompressionCodec with

```
void transform(ByteBufAllocator alloc, ByteBuf in, List<Object> out)
boolean isClosed();
void close(List<Object> out);

```
If all Encoder and Decoder implement this interface, all of them would be usable outside of a ChannelHandler and that would suit my needs for now.

On a later date, we could implement a MessageToByteEncoder and ByteToMessageDecoder that takes a CompressionCodec as parameter and remove the dependency on ChannelHandler from the compressioncodecs entirely, leaving stub classes like

```
public class Bzip2Decoder extends CompressionEncoder {
      public Bzip2Decoder(){
          super(new Bzip2Decompressor());
      }
}

```

Question is: can I build on your pull request?

normanmaurer commented:
We did fix this in main... Let me close this one
